src/osd/scrubber/osd_scrub_sched.cc

```diff
@@ -0,0 +1,724 @@
+  // scan_penalized() is called and the job was moved to the to_scrub queue.
+  updated = true;
+
+  dout(10) << " pg(" << pgid << ") adjusted: " << sched_time << "  "
+	   << registration_state() << dendl;
+}
+
+// ////////////////////////////////////////////////////////////////////////// //
+// ScrubQueue
+
+#undef dout_context
+#define dout_context (cct)
+#undef dout_prefix
+#define dout_prefix \
+  *_dout << "osd." << osd_service.whoami << " scrub-queue::" << __func__ << " "
+
+
+ScrubQueue::ScrubQueue(CephContext* cct, OSDService& osds)
+    : cct{cct}, osd_service{osds}
+{
+  // initialize the daily loadavg with current 15min loadavg
+  if (double loadavgs[3]; getloadavg(loadavgs, 3) == 3) {
+    daily_loadavg = loadavgs[2];
+  } else {
+    derr << "OSD::init() : couldn't read loadavgs\n" << dendl;
+    daily_loadavg = 1.0;
+  }
+}
+
+std::optional<double> ScrubQueue::update_load_average()
+{
+  int hb_interval = cct->_conf->osd_heartbeat_interval;
+  int n_samples = 60 * 24 * 24;
+  if (hb_interval > 1) {
+    n_samples /= hb_interval;
+    if (n_samples < 1)
+      n_samples = 1;
+  }
+
+  // get CPU load avg
+  double loadavgs[1];
+  if (getloadavg(loadavgs, 1) == 1) {
+    daily_loadavg = (daily_loadavg * (n_samples - 1) + loadavgs[0]) / n_samples;
+    dout(17) << "heartbeat: daily_loadavg " << daily_loadavg << dendl;
+    return 100 * loadavgs[0];
+  }
+
+  return std::nullopt;
+}
+
+/*
+ * Modify the scrub job state:
+ * - if 'registered' (as expected): mark as 'unregistering'. The job will be
+ *   dequeued the next time sched_scrub() is called.
+ * - if already 'not_registered': shouldn't really happen, but not a problem.
+ *   The state will not be modified.
+ * - same for 'unregistering'.
+ *
+ * Note: not holding the jobs lock
+ */
+void ScrubQueue::remove_from_osd_queue(ScrubJobRef scrub_job)
+{
+  dout(10) << "removing pg(" << scrub_job->pgid << ") from OSD scrub queue"
+	   << dendl;
+
+  qu_state_t expected_state{qu_state_t::registered};
+  auto ret = scrub_job->state.compare_exchange_strong(
+    expected_state, qu_state_t::unregistering);
+
+  if (ret) {
+
+    dout(10) << "pg(" << scrub_job->pgid << ") sched-state changed from "
+	     << qu_state_text(expected_state) << " to "
+	     << qu_state_text(scrub_job->state) << dendl;
+
+  } else {
+
+    // job wasn't in state 'registered' coming in
+    dout(5) << "removing pg(" << scrub_job->pgid
+	    << ") failed. State was: " << qu_state_text(expected_state)
+	    << dendl;
+  }
+}
+
+void ScrubQueue::register_with_osd(ScrubJobRef scrub_job,
+				   const ScrubQueue::sched_params_t& suggested)
+{
+  qu_state_t state_at_entry = scrub_job->state.load();
+
+  switch (state_at_entry) {
+    case qu_state_t::registered:
+      // just updating the schedule?
+      update_job(scrub_job, suggested);
+      break;
+
+    case qu_state_t::not_registered:
+      // insertion under lock
+      {
+	std::unique_lock lck(jobs_lock, 2'000ms);
```

Comment:
Can you explain what makes it hard to prove that this doesn't deadlock?

Reply:
I don't particularly like using a magic timed lock in production code, but I genuinely have a problem with the idea that this lock is so complicated that we're worried about further deadlocks.